wire/common.go

```diff
@@ -765,3 +835,16 @@ func isStrictAscii(s string) bool {
 
 	return true
 }
+
+// mustHash returns the hash of the serialized message.  If message
+// serialization errors, it panics with a wrapped error.
+func mustHash(msg Message, pver uint32) chainhash.Hash {
```

Comment:
This is really no different than a method like MsgTx.TxHash().  Both might panic if there's a failed serialization of the message itself.  The hashing itself can't error in either case, but only the message serialization.

Reply:
Maybe the comparison isn't totally fair.  Reading through the MsgTx serialization code, i don't see any errors that could occur that aren't originating from the writer itself.

Reply:
i've removed all of the erroring from the serialization paths that write to a hasher (for message hashing itself, and the serialization of the message contents committed to by signatures).  The only cases these might still error is if a non-hash.Hash writer is passed to WriteSigned.

we could probably use similar refactoring for transaction hashing as well, but I'm not going to bother with that here on this PR.

Reply:
(i couldn't resist looking at MsgTx a bit more)

it currently serializes to a new byte slice, and then eventually calls blake256.Sum256() instead of creating a hash.Hash object.  I recall this was done to avoid an allocation, but we're still allocating for the byte slice.

Following a similar hashing method scheme to what I currently have here on this branch now would allow that hasher to be allocated once in calling code, save on byte buffer allocations in MsgTx.serialize(), and potentially remove the blake256 dependency in wire if we do a major module bump.

something like:

```
func (tx *MsgTx) WriteHash(h hash.Hash) // uses TxSerializeNoWitness
func (tx *MsgTx) WriteHashType(h hash.Hash, serType TxSerializeType)
```

Reply:
considering this to be resolved now.  both the message hashing and the signature commitments now follow the above pattern of serializing to a hash.Hash object.  neither returns any errors in their serialization paths.  the panic in common.go is gone, and even the caller will not be tempted to panic on any hashing errors, as none are returned by these serializing methods.